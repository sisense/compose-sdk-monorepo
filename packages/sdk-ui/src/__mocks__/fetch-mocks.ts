import { MockParams } from 'vitest-fetch-mock';
import mockGlobals from './data/mock-globals.json';
import mockPalettes from './data/mock-palettes.json';
import mockWidgetDrilldown from './data/mock-widget-drilldown.json';
import mockJaqlDrilldown from './data/mock-jaql-drilldown.json';
import mockJaqlDates from './data/mock-jaql-dates.json';

export const mockUrl = `http://10.50.0.1:30845`;
export const mockToken = 'fake-token';
export const mockDashboardId = 'fake-dashboard-id';
export const mockWidgetId = 'fake-widget-id';

export const fetchMocks: Record<string, [string, MockParams]> = {
  globals: [JSON.stringify(mockGlobals), { status: 200 }],
  palettes: [JSON.stringify(mockPalettes), { status: 200 }],
  widgetDrilldown: [JSON.stringify(mockWidgetDrilldown), { status: 200 }],
  jaqlDrilldown: [JSON.stringify(mockJaqlDrilldown), { status: 200 }],
  jaqlDates: [JSON.stringify(mockJaqlDates), { status: 200 }],
};
