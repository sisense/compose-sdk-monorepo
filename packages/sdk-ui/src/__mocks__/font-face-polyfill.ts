// Polifill for FontFace class to avoid error in tests
// https://developer.mozilla.org/en-US/docs/Web/API/FontFace

export class FontFaceMock {
  family: string;

  source: string;

  descriptors: unknown;

  constructor(family: string, source: string, descriptors?: unknown) {
    this.family = family;
    this.source = source;
    this.descriptors = descriptors;
  }

  load(): Promise<FontFaceMock> {
    return Promise.resolve(this);
  }
}

globalThis.FontFace = FontFaceMock as never;
