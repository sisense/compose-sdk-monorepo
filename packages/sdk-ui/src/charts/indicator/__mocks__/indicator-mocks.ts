import { IndicatorChartData } from '@/chart-data/types';
import { IndicatorChartDesignOptions } from '@/chart-options-processor/translations/design-options';
import { IndicatorChartDataOptions } from '@/types';
import * as DM from '@/__test-helpers__/sample-ecommerce';
import { measureFactory } from '@sisense/sdk-data';

export const chartData: IndicatorChartData = {
  type: 'indicator',
  value: 14479310.424909197,
  min: 0,
  max: 125000000,
};

export const chartDesignOptions: IndicatorChartDesignOptions = {
  lineType: 'straight',
  legend: 'bottom',
  lineWidth: 2,
  valueLabel: {},
  marker: {
    enabled: false,
    size: 'small',
    fill: 'full',
  },
  xAxis: {
    type: 'linear',
    enabled: true,
    titleEnabled: true,
    title: 'X Axis title',
    gridLine: true,
    labels: true,
    min: null,
    max: null,
    tickInterval: null,
  },
  yAxis: {
    type: 'linear',
    enabled: true,
    titleEnabled: true,
    title: 'Y Axis title',
    gridLine: true,
    labels: true,
    min: null,
    max: null,
    tickInterval: null,
  },
  autoZoom: {
    enabled: true,
  },

  indicatorType: 'gauge',
  skin: 1,
  indicatorComponents: {
    title: {
      shouldBeShown: true,
      text: 'Total Revenue',
    },
    secondaryTitle: {
      text: '',
    },
    ticks: {
      shouldBeShown: true,
    },
    labels: {
      shouldBeShown: true,
    },
  },
  dataLimits: {
    seriesCapacity: 50,
    categoriesCapacity: 100000,
  },
  forceTickerView: false,
};

export const chartDataOptions: IndicatorChartDataOptions = {
  value: [measureFactory.sum(DM.Commerce.Revenue)],
  secondary: [],
  min: [measureFactory.constant(0)],
  max: [measureFactory.constant(125000000)],
};
