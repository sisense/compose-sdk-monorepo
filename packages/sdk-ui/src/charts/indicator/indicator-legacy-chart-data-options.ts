/* eslint-disable @typescript-eslint/no-use-before-define */
/* eslint-disable complexity */
/* eslint-disable max-lines-per-function */
import {
  applyFormatPlainText,
  getCompleteNumberFormatConfig,
} from '../../chart-options-processor/translations/number-format-config';
import { IndicatorChartDataOptions } from '../../chart-data-options/types';
import {
  IndicatorChartDesignOptions,
  NumericIndicatorSubType,
} from '../../chart-options-processor/translations/design-options';
import { IndicatorChartData } from '../../chart-data/types';
import { LegacyIndicatorChartTypes } from './types';
import { CompleteNumberFormatConfig, NumberFormatConfig } from '../../types';
import {
  AllowedIndicatorColoringTypes,
  getValueColorOptions,
} from './indicator-legacy-chart-options/override-with-value-color';
import {
  ColoringService,
  getColoringServiceByColorOptions,
} from '../../chart-data-options/coloring';
import { isNotAvailable } from '@/utils/not-available-value';

export type IndicatorLegacyChartDataOptions = ReturnType<typeof createLegacyChartDataOptions>;

const getValueCustomBackgroundColor = (
  chartData: IndicatorChartData,
  chartDesignOptions: IndicatorChartDesignOptions,
  chartDataOptions: IndicatorChartDataOptions,
) => {
  const colorOptions = getValueColorOptions(chartDataOptions);

  if (colorOptions && chartData.value) {
    const coloringService = getColoringServiceByColorOptions(
      colorOptions,
    ) as ColoringService<AllowedIndicatorColoringTypes>;

    return coloringService.getColor(chartData.value);
  }

  return null;
};

export const createLegacyChartDataOptions = (
  chartData: IndicatorChartData,
  chartDesignOptions: IndicatorChartDesignOptions,
  chartDataOptions: IndicatorChartDataOptions,
) => {
  const min = chartData.min || 0;
  const max = chartData.max || 100;

  const numberConfigForValue = getNumberFormatConfigForColumn(chartDataOptions, 'value');
  const numberConfigForSecondary = getNumberFormatConfigForColumn(chartDataOptions, 'secondary');
  const numberConfigForMin = getNumberFormatConfigForColumn(chartDataOptions, 'min');
  const numberConfigForMax = getNumberFormatConfigForColumn(chartDataOptions, 'max');

  const valueCustomBgColor = getValueCustomBackgroundColor(
    chartData,
    chartDesignOptions,
    chartDataOptions,
  );

  const defaultSkin = chartDesignOptions.indicatorType === 'gauge' ? 1 : 'vertical';

  return {
    ...defaultIndicatorData,
    ...(valueCustomBgColor ? { color: valueCustomBgColor } : null),
    type:
      convertToLegacyChartType(
        chartDesignOptions.indicatorType,
        'numericSubtype' in chartDesignOptions ? chartDesignOptions.numericSubtype : undefined,
      ) || 'numericSimple',
    skin: chartDesignOptions.skin || defaultSkin,
    title: { text: chartDesignOptions.indicatorComponents?.title?.text ?? '' },
    min: { data: min, text: applyFormatPlainText(numberConfigForMin, min) },
    max: { data: max, text: applyFormatPlainText(numberConfigForMax, max) },
    value: {
      data: chartData.value,
      text: getValueText(chartData.value, numberConfigForValue),
    },
    secondary: {
      data: chartData.secondary,
      text: getValueText(chartData.secondary, numberConfigForSecondary),
    },
    secondaryTitle: {
      text: chartDesignOptions.indicatorComponents?.secondaryTitle?.text ?? '',
    },
    showSecondary: chartData.secondary !== undefined,
    showTitle: !!chartDesignOptions.indicatorComponents?.title?.shouldBeShown,
    showLabels: !!chartDesignOptions.indicatorComponents?.labels?.shouldBeShown,
    showTicks: !!chartDesignOptions.indicatorComponents?.ticks?.shouldBeShown,
  };
};

function convertToLegacyChartType(
  indicatorType: IndicatorChartDesignOptions['indicatorType'],
  numericSubtype?: NumericIndicatorSubType,
): LegacyIndicatorChartTypes {
  if (indicatorType === 'numeric') {
    return numericSubtype!;
  }
  return indicatorType;
}

function getNumberFormatConfigForColumn(
  chartDataOptions: IndicatorChartDataOptions,
  columnName: keyof IndicatorChartDataOptions,
): CompleteNumberFormatConfig {
  const column = chartDataOptions[columnName]?.[0];
  return getCompleteNumberFormatConfig(
    column && 'numberFormatConfig' in column ? column.numberFormatConfig : {},
  );
}

const defaultIndicatorData = {
  color: '#00cee6',
  showSecondary: false,
};

function getValueText(
  value: undefined | string | number,
  formatConfig: Required<NumberFormatConfig>,
) {
  if (value === undefined) {
    return '';
  } else if (isNotAvailable(value)) {
    return '#N/A';
  } else {
    return applyFormatPlainText(
      formatConfig,
      typeof value === 'string' ? Number.parseFloat(value) : value,
    );
  }
}
