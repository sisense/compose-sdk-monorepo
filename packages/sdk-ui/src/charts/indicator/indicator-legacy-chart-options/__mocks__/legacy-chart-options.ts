import { NumericSimpleOptions } from '../../types';

export const numericSimpleLegacyChartOptionsWithDarkTheme: NumericSimpleOptions = {
  fontFamily: 'impact',
  title: {
    color: '#FFFFFF',
    fontSizes: {
      big: 22,
      medium: 18,
      small: 15,
      micro: 12,
    },
  },
  value: {
    fontSizes: {
      big: 66,
      medium: 46,
      small: 32,
      micro: 23,
    },
  },
  secondaryTitle: {
    color: 'rgb(163, 163, 163)',
    fontSizes: {
      big: 20,
      medium: 14,
      small: 10,
      micro: 10,
    },
  },
  secondaryValue: {
    fontWeight: 800,
    color: 'rgb(163, 163, 163)',
    fontSizes: {
      big: 20,
      medium: 14,
      small: 10,
      micro: 10,
    },
  },
  textKeys: ['title', 'value', 'secondaryTitle', 'secondaryValue'],
  backgroundColor: '#333333',
  borderColor: 'rgb(215, 215, 215)',
  borderWidth: 1,
  relativeSizes: [
    {
      key: 'titleSectionHeight',
      value: 0.5,
      decimals: 0,
    },
    {
      key: 'titleSectionMarginTop',
      value: 0.13,
      decimals: 0,
    },
    {
      key: 'titleSectionMinWidth',
      value: 0.29,
      decimals: 0,
    },
    {
      key: 'numericMinWidth',
      value: 4.69,
      decimals: 0,
    },
    {
      key: 'valueSectionHeight',
      value: 1,
      decimals: 0,
    },
    {
      key: 'valueSectionMarginRight',
      value: 0.14,
      decimals: 0,
    },
    {
      key: 'valueSectionMarginBottom',
      value: 0.07,
      decimals: 0,
    },
    {
      key: 'secSectionHeight',
      value: 0.34,
      decimals: 0,
    },
    {
      key: 'secSectionMarginTop',
      value: 0.14,
      decimals: 0,
    },
    {
      key: 'secDividerWidth',
      value: 0.14,
      decimals: 0,
    },
    {
      key: 'secTitleMinWidth',
      value: 0.29,
      decimals: 0,
    },
    {
      key: 'indicatorHorizontalMargin',
      value: 0.31,
      decimals: 0,
    },
    {
      key: 'indicatorVerticalMargin',
      value: 0.46,
      decimals: 0,
    },
  ],
  measureKeys: [
    'indicatorMargin',
    'numericMinWidth',
    'titleSectionHeight',
    'secSectionHeight',
    'valueSectionMarginBottom',
    'secSectionMarginTop',
    'indicatorHorizontalMargin',
    'indicatorVerticalMargin',
  ],
  sizes: [
    {
      size: 'big',
      value: 70,
    },
    {
      size: 'medium',
      value: 49,
    },
    {
      size: 'small',
      value: 34,
    },
    {
      size: 'micro',
      value: 24,
    },
  ],
  fitMeasures: [
    {
      key: 'fitTitleMeasure',
      prop: 'string',
      dataKey: 'title',
    },
    {
      key: 'fitValueMeasure',
      prop: 'string',
      dataKey: 'value',
    },
    {
      key: 'fitSecondaryMeasure',
      prop: 'titleString',
      dataKey: 'secondaryTitle',
    },
    {
      key: 'fitSecondaryMeasure',
      prop: 'valueString',
      dataKey: 'secondary',
    },
  ],
};
