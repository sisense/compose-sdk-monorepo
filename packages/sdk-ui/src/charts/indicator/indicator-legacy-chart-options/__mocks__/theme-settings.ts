import { ThemeSettings } from '../../../../types.js';

export const darkThemeSettings: ThemeSettings = {
  chart: {
    backgroundColor: '#333333',
    textColor: '#FFFFFF',
    panelBackgroundColor: '#F6F6F6',
  },
  typography: {
    fontFamily: 'impact',
  },
  palette: {
    variantColors: ['#00cee6', '#9b9bd7', '#6eda55', '#fc7570', '#fbb755', '#218a8c'],
  },
  general: {
    backgroundColor: '#ffffff',
    brandColor: '#ffcb05',
    primaryButtonTextColor: '#3a4356',
  },
};
