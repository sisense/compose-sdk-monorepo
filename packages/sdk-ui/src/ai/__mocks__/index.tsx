import { SisenseContextProvider } from '@/sisense-context/sisense-context-provider';
import { mockToken, mockUrl } from '@/__mocks__/msw';
import { QueryClient, QueryClientProvider } from '@tanstack/react-query';
import { ReactNode } from 'react';
import AiContextProvider from '../ai-context-provider';

/**
 * A component that wraps an AI hook/component with all required providers.
 * For use in unit tests only.
 *
 * @internal
 */
export const AiTestWrapper = ({ children }: { children: ReactNode }) => (
  <SisenseContextProvider url={mockUrl} token={mockToken}>
    <AiContextProvider>
      <QueryClientProvider
        client={
          new QueryClient({
            defaultOptions: {
              queries: {
                retry: false,
              },
            },
          })
        }
      >
        {children}
      </QueryClientProvider>
    </AiContextProvider>
  </SisenseContextProvider>
);
