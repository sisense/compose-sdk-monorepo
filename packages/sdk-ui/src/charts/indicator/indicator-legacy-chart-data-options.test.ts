import { MeasureColumn } from '@ethings-os/sdk-data';
import { createLegacyChartDataOptions } from './indicator-legacy-chart-data-options';
import { chartData, chartDataOptions, chartDesignOptions } from './__mocks__/indicator-mocks';

describe('indicator-legacy-chart-data-options', () => {
  describe('createLegacyChartDataOptions', () => {
    it('should return the correct legacy chart data options', () => {
      const legacyChartDataOptions = createLegacyChartDataOptions(
        chartData,
        chartDesignOptions,
        chartDataOptions,
      );
      expect(legacyChartDataOptions).toMatchSnapshot();
    });

    it('should return the correct legacy chart data options when value defined as StyledMeasureColumn', () => {
      const legacyChartDataOptions = createLegacyChartDataOptions(chartData, chartDesignOptions, {
        ...chartDataOptions,
        value: [
          {
            column: chartDataOptions.value![0] as MeasureColumn,
            color: '#FF0000',
          },
        ],
      });
      expect(legacyChartDataOptions).toMatchSnapshot();
    });

    it('should return the correct legacy chart data options when value is "N\\A"', () => {
      const legacyChartDataOptions = createLegacyChartDataOptions(
        { ...chartData, value: NaN },
        chartDesignOptions,
        chartDataOptions,
      );
      expect(legacyChartDataOptions).toMatchSnapshot();
    });
  });
});
