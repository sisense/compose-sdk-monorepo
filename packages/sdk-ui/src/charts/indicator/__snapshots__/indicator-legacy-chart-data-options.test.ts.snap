// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`indicator-legacy-chart-data-options > createLegacyChartDataOptions > should return the correct legacy chart data options 1`] = `
{
  "color": "#00cee6",
  "max": {
    "data": 125000000,
    "text": "125M",
  },
  "min": {
    "data": 0,
    "text": "0",
  },
  "secondary": {
    "data": undefined,
    "text": "",
  },
  "secondaryTitle": {
    "text": "",
  },
  "showLabels": true,
  "showSecondary": false,
  "showTicks": true,
  "showTitle": true,
  "skin": 1,
  "title": {
    "text": "Total Revenue",
  },
  "type": "gauge",
  "value": {
    "data": 14479310.424909197,
    "text": "14.48M",
  },
}
`;

exports[`indicator-legacy-chart-data-options > createLegacyChartDataOptions > should return the correct legacy chart data options when value defined as StyledMeasureColumn 1`] = `
{
  "color": "#FF0000",
  "max": {
    "data": 125000000,
    "text": "125M",
  },
  "min": {
    "data": 0,
    "text": "0",
  },
  "secondary": {
    "data": undefined,
    "text": "",
  },
  "secondaryTitle": {
    "text": "",
  },
  "showLabels": true,
  "showSecondary": false,
  "showTicks": true,
  "showTitle": true,
  "skin": 1,
  "title": {
    "text": "Total Revenue",
  },
  "type": "gauge",
  "value": {
    "data": 14479310.424909197,
    "text": "14.48M",
  },
}
`;

exports[`indicator-legacy-chart-data-options > createLegacyChartDataOptions > should return the correct legacy chart data options when value is "N\\A" 1`] = `
{
  "color": "#00cee6",
  "max": {
    "data": 125000000,
    "text": "125M",
  },
  "min": {
    "data": 0,
    "text": "0",
  },
  "secondary": {
    "data": undefined,
    "text": "",
  },
  "secondaryTitle": {
    "text": "",
  },
  "showLabels": true,
  "showSecondary": false,
  "showTicks": true,
  "showTitle": true,
  "skin": 1,
  "title": {
    "text": "Total Revenue",
  },
  "type": "gauge",
  "value": {
    "data": NaN,
    "text": "#N/A",
  },
}
`;
